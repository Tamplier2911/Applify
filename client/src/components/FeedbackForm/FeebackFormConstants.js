const feedbackFormData = {
  eng: {
    feedbackFormTitle: [
      "Please, describe what did you like about my services, and what you didn't.",
      "How could I improve?",
    ],
    feedbackFormFeedback: "Feedback",
    feedbackFormSubmit: "Submit",
    feedbackFormOptions: [
      { name: "Awful", value: "1" },
      { name: "Could be better...", value: "2" },
      { name: "Good enough", value: "3" },
      { name: "Meet expectations", value: "4" },
      { name: "Unforgettable", value: "5" },
    ],
    feedbackFormLength: "Feedback length:",
  },
  rus: {
    feedbackFormTitle: [
      "Поделитесь вашим опытом использования моих услуг.",
      "Над чем стоит поработать?",
    ],
    feedbackFormFeedback: "Отзыв",
    feedbackFormSubmit: "Разместить",
    feedbackFormOptions: [
      { name: "Ужасно", value: "1" },
      { name: "Могло быть лучше...", value: "2" },
      { name: "Достаточно хорошо", value: "3" },
      { name: "Оправдывает ожидания", value: "4" },
      { name: "Незабываемо", value: "5" },
    ],
    feedbackFormLength: "Длина отзыва:",
  },
  ukr: {
    feedbackFormTitle: [
      "Поділіться вашим досвідом використання моїх послуг.",
      "Над чим варто попрацювати?",
    ],
    feedbackFormFeedback: "Відгук",
    feedbackFormSubmit: "Розмістити",
    feedbackFormOptions: [
      { name: "Жахливо", value: "1" },
      { name: "Могло бути краще...", value: "2" },
      { name: "Достатньо добре", value: "3" },
      { name: "Виправдовує очікування", value: "4" },
      { name: "Незабутньо", value: "5" },
    ],
    feedbackFormLength: "Довжина відгука:",
  },
};

export default feedbackFormData;
