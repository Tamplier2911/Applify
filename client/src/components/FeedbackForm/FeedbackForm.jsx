// import "./FeedbackForm.scss";
import React, { useState } from "react";

// redux
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectCurrentLanguage } from "../../redux/lang/lang.selectors";
import {
  sendFeedbackStart,
  updateFeedbackStart,
} from "../../redux/feedbacks/feedbacks.actions";
import { openModal } from "../../redux/modal/modal.actions";

// components
import TextInput from "../TextInput/TextInput";
import SelectInput from "../SelectInput/SelectInput";
import Button from "../Button/Button";

// JS Rendering CSS
import {
  FeedbackFormContainer,
  FeedbackFormTitle,
  FeedbackFormParagraph,
  FeedbackFormElement,
  FeedbackFormRatingSVG,
  FeedbackFormLength,
} from "./FeedbackFormStyles";

// component constants
import feedbackFormData from "./FeebackFormConstants";

export const FeedbackForm = ({
  lang,
  openModal,
  sendFeedbackStart,
  updateFeedbackStart,
  method,
  updateData,
}) => {
  const [feedback, setFeedback] = useState({
    feedbackMessage: method === "POST" ? "" : updateData.feedback,
    feedbackRating: method === "POST" ? "5" : updateData.rating,
    feedbackId: method === "POST" ? undefined : updateData._id,
  });

  const { feedbackMessage, feedbackRating } = feedback;

  const onInputChange = (e) => {
    const { name, value } = e.target;
    setFeedback({ ...feedback, [name]: value });
  };

  const onSubmit = (e) => {
    e.preventDefault();
    if (feedbackMessage.length > 500)
      return openModal({
        header: "Attention!",
        content: "Feedback must not consist of more than 500 characters.",
      });
    if (feedbackRating.length > 1 || typeof feedbackRating !== "string")
      return openModal({
        header: "Attention!",
        content:
          "Rating must be defined as a string, rating must not consist of more than one character.",
      });
    if (!feedbackMessage.length || !feedbackRating.length)
      return openModal({
        header: "Attention!",
        content:
          "Please, enter your feedback. Use select bar to choose rating.",
      });
    method === "POST"
      ? sendFeedbackStart(feedback) &&
        setFeedback({ feedbackMessage: "", feedbackRating: "5" })
      : updateFeedbackStart(feedback);
  };

  const labels = Array.from(new Array(Number(feedbackRating)), (n, i) => i + 1);

  const {
    feedbackFormTitle,
    feedbackFormFeedback,
    feedbackFormSubmit,
    feedbackFormOptions,
    feedbackFormLength,
  } = feedbackFormData[lang];

  return (
    <FeedbackFormContainer>
      <FeedbackFormTitle lang={lang}>
        <FeedbackFormParagraph>{feedbackFormTitle[0]}</FeedbackFormParagraph>
        <FeedbackFormParagraph>{feedbackFormTitle[1]}</FeedbackFormParagraph>
      </FeedbackFormTitle>
      <FeedbackFormElement autoComplete="off" onSubmit={(e) => onSubmit(e)}>
        <TextInput
          onInputChange={(e) => onInputChange(e)}
          value={feedbackMessage}
          name="feedbackMessage"
          label={feedbackFormFeedback}
          rows="6"
          max="500"
          //   placeholder="Please, leave me a feedback!"
          required
        />
        <SelectInput
          onInputChange={(e) => onInputChange(e)}
          defaultValue={method === "POST" ? "5" : feedbackRating}
          name={"feedbackRating"}
          label={labels.map((label) => (
            <FeedbackFormRatingSVG key={label} />
          ))}
          options={feedbackFormOptions}
          required
        />
        <Button type="submit" value={feedbackFormSubmit} />
      </FeedbackFormElement>
      <FeedbackFormLength>
        {feedbackFormLength + " "}
        {500 - feedbackMessage.length}
      </FeedbackFormLength>
    </FeedbackFormContainer>
  );
};

const mapStateToProps = createStructuredSelector({
  lang: selectCurrentLanguage,
});

export default connect(mapStateToProps, {
  sendFeedbackStart,
  updateFeedbackStart,
  openModal,
})(FeedbackForm);
