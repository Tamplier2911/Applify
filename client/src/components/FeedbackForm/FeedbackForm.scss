/*

.feedback-form {
  display: grid;
  grid-row-gap: 4rem;
  grid-auto-rows: min-content;

  &__title {
    // ${getFont}
    font-weight: 400;
    font-size: 1.7rem;

    // remove that when font is getten
    font-family: var(--font-title);
  }

  &__form {
    display: grid;
    grid-row-gap: 2rem;
  }

  &__svg {
    width: 3rem;
    height: 3rem;
  }
}

*/
