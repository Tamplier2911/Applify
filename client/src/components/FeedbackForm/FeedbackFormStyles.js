import styled from "styled-components";

// svg
import { ReactComponent as RatingSVG } from "../../assets/svg/star.svg";

const getFont = props => {
  const { lang } = props;
  if (lang === "eng") {
    return `font-family: var(--font-title);`;
  } else if (lang === "rus" || lang === "ukr") {
    return `font-family: var(--font-cyrillic);`;
  }
  return `font-family: var(--font-title);`;
};

export const FeedbackFormContainer = styled.div`
  display: grid;
  grid-row-gap: 4rem;
  grid-auto-rows: min-content;
`;

export const FeedbackFormTitle = styled.h3`
  ${getFont}
  font-weight: 400;
  font-size: 1.7rem;
`;

export const FeedbackFormParagraph = styled.p``;

export const FeedbackFormElement = styled.form`
  display: grid;
  grid-row-gap: 2rem;
`;

export const FeedbackFormRatingSVG = styled(RatingSVG)`
  width: 3rem;
  height: 3rem;

  @media only screen and (max-width: 310px) {
    width: 1.5rem;
    height: 1.5rem;
  }
`;

export const FeedbackFormLength = styled.p``;
