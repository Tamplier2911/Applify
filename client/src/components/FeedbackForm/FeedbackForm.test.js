import React from "react";
import toJson from "enzyme-to-json";

// component
import FeedbackForm from "./FeedbackForm";

// utils
import {
  setShallowC,
  findByTestAttr,
  checkProps,
  storeFactory,
} from "../../tests/testUtils";

describe("<FeedbackForm />", () => {
  const initialStore = {
    lang: {
      languagePanelHidden: true,
      currentLanguage: "eng",
    },
  };

  const store = storeFactory(initialStore);

  let wrapper;
  beforeEach(() => {
    wrapper = setShallowC(<FeedbackForm store={store} method="POST" />);
  });

  it("renders without an error", () => {
    const feedbackForm = findByTestAttr(wrapper, "feedback-form");
    expect(feedbackForm.length).toBe(1);
  });

  it("renders children - feedback form", () => {
    const feedbackFormElement = findByTestAttr(
      wrapper,
      "feedback-form-element"
    );
    expect(feedbackFormElement.length).toBe(1);
  });

  it("renders children - feedback inputs", () => {
    const feedbackFormInputs = findByTestAttr(wrapper, "feedback-form-input");
    expect(feedbackFormInputs.length).toBe(2);
  });

  it("recieves expeted props", () => {
    const expectedProps = {
      method: "POST",
      updateData: { feedback: "feed", rating: "5", _id: "123213" },
    };
    const err = checkProps(FeedbackForm, expectedProps, "props");
    expect(err).toBeUndefined();
  });

  it("matches snapshot", () => {
    const feedbackForm = findByTestAttr(wrapper, "feedback-form");
    expect(toJson(feedbackForm)).toMatchSnapshot();
  });
});
