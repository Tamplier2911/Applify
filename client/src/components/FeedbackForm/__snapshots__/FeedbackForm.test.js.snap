// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeedbackForm /> matches snapshot 1`] = `
<styled.div
  data-test="feedback-form"
>
  <styled.h3
    lang="eng"
  >
    <styled.p>
      Please, describe what did you like about our services, and what you didn't.
    </styled.p>
    <styled.p>
      How could we improve?
    </styled.p>
  </styled.h3>
  <styled.form
    autoComplete="off"
    data-test="feedback-form-element"
    onSubmit={[Function]}
  >
    <TextInput
      data-test="feedback-form-input"
      label="Feedback"
      max="500"
      name="feedbackMessage"
      onInputChange={[Function]}
      required={true}
      rows="6"
      value=""
    />
    <SelectInput
      data-test="feedback-form-input"
      defaultValue="5"
      label={
        Array [
          <ForwardRef(Styled(Component)) />,
          <ForwardRef(Styled(Component)) />,
          <ForwardRef(Styled(Component)) />,
          <ForwardRef(Styled(Component)) />,
          <ForwardRef(Styled(Component)) />,
        ]
      }
      name="feedbackRating"
      onInputChange={[Function]}
      options={
        Array [
          Object {
            "name": "Awful",
            "value": "1",
          },
          Object {
            "name": "Could be better...",
            "value": "2",
          },
          Object {
            "name": "Good enough",
            "value": "3",
          },
          Object {
            "name": "Meet expectations",
            "value": "4",
          },
          Object {
            "name": "Unforgettable",
            "value": "5",
          },
        ]
      }
      required={true}
    />
    <Button
      type="submit"
      value="Submit"
    />
  </styled.form>
  <styled.p>
    Feedback length: 
    500
  </styled.p>
</styled.div>
`;
